### DISCUSSION TITLE
Trying to add RFID Door Panel from Scout

### DISCUSSION THREAD

**Discussion by hemandk** [Koenkk/zigbee2mqtt#12665] [Devices]
Hi,

So I bought a door panel (RFID reader with buzzer etc) from Scout Alarm and I trying to add it as a new device, but I'm kinda lost to be honest.

I can scan a tag and I can see tag ID in the log:
```
Zigbee2MQTT:debug 2022-05-31 23:52:32: Received Zigbee message from 'ScoutDoorPanel', type 'raw', cluster 'manuSpecificUbisysDimmerSetup', data '{"data":[5,2,16,16,0,48,51,48,48,52,55,55,48,50,49],"type":"Buffer"}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2022-05-31 23:52:32: No converter available for 'SC-EDP-01' with cluster 'manuSpecificUbisysDimmerSetup' and type 'raw' and data '{"data":[5,2,16,16,0,48,51,48,48,52,55,55,48,50,49],"type":"Buffer"}'
```

My converter so far:
```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['Door Panel 2'], // The model ID from: Device with modelID 'lumi.sens' is not supported.
    model: 'SC-EDP-01', // Vendor model number, look on the device for a model number
    vendor: 'Scout', // Vendor of the device (only used for documentation and startup logging)
    description: 'Alarm panel with RFID', // Description of the device, copy from vendor site. (only used for documentation and startup logging)
    fromZigbee: [], // We will add this later
    toZigbee: [], // Should be empty, unless device can be controlled (e.g. lights, switches).
    exposes: [e.temperature(), e.humidity()], // Defines what this device exposes, used for e.g. Home Assistant discovery and in the frontend
};

module.exports = definition;
```
I can read the temperature and humidity clusters and get a value, but nothing shows in the zigbee2mqtt dashboard.

I managed to get some info from the supplier and I can properly get more. They have been very helpfull:

the RFID custom cluster id is 0xFC01
the RFID string is sent on command id 0x00
our Scout Inc zigbee manufacturer id is 0x1202
the payload of the command will contain the RFID string in ASCII hex format
the panel is the server and the controller the client
the client can send a 'code is valid / not valid' on command id 0x01 and the panel LED will reflect it
the payload for valid is 0x01 and not valid is 0x00
the panel also exposes some standard zigbee clusters -
power (0x0001) to read the battery voltage (0x0020)
temperature (0x0402) to read temperature (0x0000)
humidity (0x0405) to read humidity (0x0000)
IAS WD (0x0502) to sound the siren and chirp

So can anybody help me with making a better converter?

**Comment by sjorge** [comment]
Some good news, the provided from the manufacturer should be enough to get it working, the bad news, it's gonna require a lot of trial and error.

I can nudge you in the right direction though but without the device it gonna be hard to guide someone to add support.

> the RFID custom cluster id is 0xFC01 the RFID string is sent on command id 0x00

```
Zigbee2MQTT:debug 2022-05-31 23:52:32: Received Zigbee message from 'ScoutDoorPanel', type 'raw', cluster 'manuSpecificUbisysDimmerSetup', data '{"data":[5,2,16,16,0,48,51,48,48,52,55,55,48,50,49],"type":"Buffer"}' from endpoint 1 with groupID 0
```

OK so thats this! We're getting the data. Why does it show manuSpecificUbisysDimmerSetup? Because that's the first match for `0xFC01` which is the custom cluster ID range from the ZCL.

Once the right definition is added to zigbee-herdsman it should show up properly and then you can continue with sending the reply back to the device based on the RFID string.

So the first step is getting a definition into herdsman, it should be added here : https://github.com/Koenkk/zigbee-herdsman/blob/5631a0e531ffaa10f2a5d224d1fb62e04cd81988/src/zcl/definition/cluster.ts#L4354

Something like this, tweaking may be required:
```
    manuSpecificScoutSecurity: {
        ID: 0xfc01,
        manufacturerCode: ManufacturerCode.SCOUT_SECURITY,
        attributes: {
            rfid: {ID: 0x00, type: DataType.charStr},
        },
        commands: {
            sendResponse: {ID: 0x01, type: DataType.boolean},
        },
        commandsResponse: {
        }
    },
```

You might need to play with `charStr` and try some other varients https://github.com/Koenkk/zigbee-herdsman/blob/5631a0e531ffaa10f2a5d224d1fb62e04cd81988/src/zcl/definition/dataType.ts#L37-L40

Same for the `boolean` on the sendReponse command, it might be something else too I'm just doing my best guess based on the info you provided.


Once that is in place, you should hopefully see `No converter available for 'SC-EDP-01' with cluster manuSpecificScoutSecurity ...`

**Reply by hemandk** [reply]
Thanks for the reply.

So this is in no way my area of expertise, but can I just edit:
https://github.com/Koenkk/zigbee-herdsman/blob/5631a0e531ffaa10f2a5d224d1fb62e04cd81988/src/zcl/definition/cluster.ts#L4354 
and add your suggestion or how does it work?

**Reply by sjorge** [reply]
You'd need to clone the repo, make the change and then install zigbee-herdsman from you're repo. If it works, you can then open a PR against upstream to get it merged.

**Reply by sjorge** [reply]
@Koenkk do we have a more indepth dev guide somewhere on how to run your own fork of herdsman? I thought we did but can't find it.
